File: src/ros/time.cpp
#include "torch-ros.h"

ROSIMP(ros::Time*, Time, new)() {
  return new ros::Time();
}

ROSIMP(void, Time, delete)(ros::Time *self) {
  delete self;
}

ROSIMP(ros::Time*, Time, clone)(ros::Time *self) {
  return new ros::Time(*self);
}

ROSIMP(bool, Time, isZero)(ros::Time *self) {
  return self->isZero();
}

ROSIMP(void, Time, fromSec)(ros::Time *self, double t) {
  self->fromSec(t);
}

ROSIMP(double, Time, toSec)(ros::Time *self) {
  return self->toSec();
}

ROSIMP(void, Time, set)(ros::Time *self, unsigned int sec, unsigned int nsec) {
  self->sec = sec;
  self->nsec = nsec;
}

ROSIMP(void, Time, assign)(ros::Time *self, ros::Time *other) {
  *self = *other;
}

ROSIMP(int, Time, get_sec)(ros::Time *self) {
  return static_cast<int>(self->sec);
}

ROSIMP(void, Time, set_sec)(ros::Time *self, unsigned int sec) {
  self->sec = sec;
}

ROSIMP(int, Time, get_nsec)(ros::Time *self) {
  return static_cast<int>(self->nsec);
}

ROSIMP(void, Time, set_nesc)(ros::Time *self, unsigned int nsec) {
  self->nsec = nsec;
}

ROSIMP(bool, Time, lt)(ros::Time *self, ros::Time *other) {
  return self->operator<(*other);
}

ROSIMP(bool, Time, eq)(ros::Time *self, ros::Time *other) {
  return self->operator==(*other);
}

ROSIMP(void, Time, add_Duration)(ros::Time *self, ros::Duration *duration, ros::Time *result) {
  *result = self->operator+(*duration);
}

ROSIMP(void, Time, sub)(ros::Time *self, ros::Time *other, ros::Duration *result) {
  *result = self->operator-(*other);
}

ROSIMP(void, Time, sub_Duration)(ros::Time *self, ros::Duration *duration, ros::Time *result) {
  *result = self->operator-(*duration);
}

// static members

ROSIMP(void, Time, sleepUntil)(ros::Time *end) {
  ros::Time::sleepUntil(*end);
}

ROSIMP(void, Time, getNow)(ros::Time *result) {
  *result = ros::Time::now();
}

ROSIMP(void, Time, setNow)(ros::Time* now) {
  ros::Time::setNow(*now);
}
    
ROSIMP(void, Time, waitForValid)() {
  ros::Time::waitForValid();
}

ROSIMP(void, Time, init)() {
  ros::Time::init();
}

ROSIMP(void, Time, shutdown)() {
  ros::Time::shutdown();
}

ROSIMP(bool, Time, useSystemTime)() {
  return ros::Time::useSystemTime();
}

ROSIMP(bool, Time, isSimTime)() {
  return ros::Time::isSimTime();
}

ROSIMP(bool, Time, isSystemTime)() {
  return ros::Time::isSystemTime();
}

ROSIMP(bool, Time, isValid)() {
  return ros::Time::isValid();
}

File: lua/Time.lua
local ffi = require 'ffi'
local torch = require 'torch'
local ros = require 'ros.env'
local utils = require 'ros.utils'

local Time = torch.class('ros.Time', ros)

function init()
  local Time_method_names = {
    "new",
    "clone",
    "delete",
    "isZero",
    "fromSec",
    "toSec",
    "set",
    "assign",
    "get_sec",
    "set_sec",
    "get_nsec",
    "set_nesc",
    "lt",
    "eq",
    "add_Duration",
    "sub",
    "sub_Duration",
    "sleepUntil",
    "getNow",
    "setNow",
    "waitForValid",
    "init",
    "shutdown",
    "useSystemTime",
    "isSimTime",
    "isSystemTime",
    "isValid"
  }

  return utils.create_method_table("ros_Time_", Time_method_names)
end

local f = init()

function Time:__init(_1, _2)
  self.o = f.new()
  if _1 or _2 then
    self:set(_1, _2)
  end
end

function Time:cdata()
  return self.o
end

function Time:clone()
  local c = torch.factory('ros.Time')()
  rawset(c, 'o', f.clone(self.o))
  return c
end

function Time:isZero()
  return f.isZero(self.o)
end

function Time:fromSec(t)
  f.fromSec(self.o, t)
end

function Time:toSec()
  return f.toSec(self.o)
end

function Time:set(_1, _2)
  if torch.isTypeOf(_1, ros.Time) then
    self:assign(_1)
  elseif not _2 then
    self:fromSec(_1)
  else
    f.set(self.o, _1, _2)
  end
end

function Time:get()
  return self:get_sec(), self:get_nsec()
end

function Time:assign(other)
  f.assign(self.o, other:cdata())
end

function Time:get_sec()
  return f.get_sec(self.o)
end

function Time:set_sec(sec)
  f.set_sec(self.o, sec)
end

function Time:get_nsec()
  return f.get_nsec(self.o)
end

function Time:set_nesc(nsec)
  f.set_nesc(self.o, nsec)
end

function Time:__lt(other)
  return f.lt(self.o, other:cdata())
end

function Time:__eq(other)
  return f.eq(self.o, other:cdata())
end

function Time:add(d, result)
  result = result or self
  if type(d) == 'number' then
    d = ros.Duration(d)
  end
  f.add_Duration(self.o, d:cdata(), result:cdata())
  return result
end

function Time:sub(x, result)
  if type(x) == 'number' then
    x = ros.Duration(x)
  end
  if torch.isTypeOf(x, ros.Time) then
    result = result or ros.Duration()
    f.sub(self.o, x:cdata(), result:cdata())
  elseif torch.isTypeOf(x, ros.Duration) then
    result = result or self
    f.sub_Duration(self.o, x:cdata(), result:cdata())
  else
    error('cannot sub from ros.Time with specified argument type')
  end
  return result
end

function Time:__add(d)
  local result = ros.Time()
  return self:add(d, result)
end

function Time:__sub(x)
  local result
  if torch.isTypeOf(x, ros.Time) then
    result = ros.Duration()
    f.sub(self.o, x:cdata(), result:cdata())
  elseif torch.isTypeOf(x, ros.Duration) then
    result = ros.Time()
    f.sub_Duration(self.o, x:Duration(), result:cdata())
  end
  return result
end

function Time:__tostring()
  return string.format("%f", self:toSec())
end

-- static functions

function Time.init()
  f.init()
end

function Time.shutdown()
  f.shutdown()
end

function Time.now()
  return Time.getNow()
end

function Time.getNow(result)
  result = result or ros.Time()
  f.getNow(result:cdata())
  return result
end

function Time.setNow(time)
  f.setNow(time:cdata())
end

function Time.sleepUntil(time)
  f.sleepUntil(time:cdata())
end

function Time.waitForValid()
  f.waitForValid()
end

function Time.useSystemTime()
  return f.useSystemTime()
end

function Time.isSimTime()
  return f.isSimTime()
end

function Time.isSystemTime()
  return f.isSimTime()
end

function Time.isValid()
  return f.isValid()
end

Time.init()
